tests/bluetooth/bsim_bt/bsim_test_l2cap/src/main_l2cap_ecred.c

```diff
@@ -0,0 +1,618 @@
+/* main_l2cap_ecred.c - Application main entry point */
+
+/*
+ * Copyright (c) 2022 Nordic Semiconductor
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "common.h"
+
+#define LOG_MODULE_NAME main_l2cap_ecred
+#include <logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME, LOG_LEVEL_DBG);
+
+extern enum bst_result_t bst_result;
+
+static struct bt_conn *default_conn;
+
+static const struct bt_data ad[] = {
+	BT_DATA_BYTES(BT_DATA_FLAGS, (BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR)),
+};
+
+#define DATA_MTU CONFIG_BT_L2CAP_TX_MTU
+#define DATA_MPS 65
+#define DATA_BUF_SIZE BT_L2CAP_SDU_BUF_SIZE(DATA_MTU)
+#define L2CAP_CHANNELS 2
+#define SERVERS 1
+#define SDU_SEND_COUNT 200
+#define ECRED_CHAN_MAX 5
+#define LONG_MSG (DATA_MTU - 500)
+#define SHORT_MSG (DATA_MPS - 2)
+#define LONG_MSG_CHAN_IDX 0
+#define SHORT_MSG_CHAN_IDX 1
+
+NET_BUF_POOL_FIXED_DEFINE(rx_data_pool, L2CAP_CHANNELS, BT_L2CAP_BUF_SIZE(DATA_BUF_SIZE), 8, NULL);
+NET_BUF_POOL_FIXED_DEFINE(tx_data_pool_0, 1, BT_L2CAP_BUF_SIZE(DATA_MTU), 8, NULL);
+NET_BUF_POOL_FIXED_DEFINE(tx_data_pool_1, 1, BT_L2CAP_BUF_SIZE(DATA_MTU), 8, NULL);
+
+static struct bt_l2cap_server servers[SERVERS];
+void send_sdu_chan_worker(struct k_work *item);
+struct channel {
+	uint8_t chan_id; /* Internal number that identifies L2CAP channel. */
+	struct bt_l2cap_le_chan le;
+	bool in_use;
+	int sdus_received;
+	int bytes_to_send;
+	int iteration;
+	struct net_buf *buf;
+	struct k_work work;
+	struct k_work_q work_queue;
+	uint8_t payload[DATA_MTU];
+};
+static struct channel channels[L2CAP_CHANNELS];
+
+CREATE_FLAG(is_connected);
+CREATE_FLAG(unsequenced_data);
+
+#define T_STACK_SIZE 512
+#define T_PRIORITY 5
+
+static K_THREAD_STACK_ARRAY_DEFINE(stack_area, L2CAP_CHANNELS, T_STACK_SIZE);
+static K_SEM_DEFINE(chan_conn_sem, 0, L2CAP_CHANNELS);
+static K_SEM_DEFINE(all_chan_conn_sem, 0, 1);
+static K_SEM_DEFINE(all_chan_disconn_sem, 0, 1);
+static K_SEM_DEFINE(sent_sem, 0, L2CAP_CHANNELS);
+
+static void init_workqs(void)
+{
+	for (int i = 0; i < L2CAP_CHANNELS; i++) {
+		k_work_queue_init(&channels[i].work_queue);
+		k_work_queue_start(&channels[i].work_queue, stack_area[i],
+		K_THREAD_STACK_SIZEOF(*stack_area), T_PRIORITY, NULL);
+	}
+}
+
+static struct net_buf *chan_alloc_buf_cb(struct bt_l2cap_chan *chan)
+{
+	LOG_DBG("Allocated on chan %p", chan);
+	return net_buf_alloc(&rx_data_pool, K_FOREVER);
+}
+
+static int chan_recv_cb(struct bt_l2cap_chan *l2cap_chan, struct net_buf *buf)
+{
+	struct channel *chan = CONTAINER_OF(l2cap_chan, struct channel, le);
+	uint32_t received_iterration = net_buf_pull_le32(buf);
+
+	LOG_DBG("received_iterration %i sdus_receied %i, chan_id: %d, data_length: %d",
+		received_iterration, chan->sdus_received, chan->chan_id, buf->len);
+	if (!unsequenced_data && received_iterration != chan->sdus_received) {
```

Comment:
Consider adding something like 
```
#define TEST_FLAG(flag) (atomic_get(&flag) == (atomic_t)true)
``` 

and do `TEST_FLAG(unsequenced_data)` to do a proper comparison of an atomic_t value.